core/src/Kokkos_View.hpp

```diff
@@ -19,13 +19,1008 @@
+                 static_cast<pointer_type>(arg_space.get_shmem_aligned(
+                     required_allocation_size(typename traits::array_layout(
+                         arg_N0, arg_N1, arg_N2, arg_N3, arg_N4, arg_N5, arg_N6,
+                         arg_N7)),
+                     scratch_value_alignment))),
+             typename traits::array_layout(arg_N0, arg_N1, arg_N2, arg_N3,
+                                           arg_N4, arg_N5, arg_N6, arg_N7)) {
+    static_assert(traits::array_layout::is_extent_constructible,
+                  "Layout is not constructible from extent arguments. Use "
+                  "overload taking a layout object instead.");
+  }
+
+ public:
+  //----------------------------------------
+  // Allocation tracking properties
+  std::string label() const {
+    if constexpr (traits::is_managed) {
+      return this->data_handle().get_label();
+    } else {
+      return "";
+    }
+  }
+
+  int use_count() const {
+    if constexpr (traits::is_managed) {
+      return this->data_handle().use_count();
+    } else {
+      return 0;
+    }
+  }
+
+  KOKKOS_FUNCTION
+  constexpr typename base_t::index_type extent(size_t r) const noexcept {
+    // casting to int to avoid warning for pointless comparison of unsigned
+    // with 0
+    if (static_cast<int>(r) >= static_cast<int>(base_t::extents_type::rank()))
+      return 1;
+    return base_t::extent(r);
+  }
+  KOKKOS_FUNCTION
+  static constexpr size_t static_extent(size_t r) noexcept {
+    // casting to int to avoid warning for pointless comparison of unsigned
+    // with 0
+    if (static_cast<int>(r) >= static_cast<int>(base_t::extents_type::rank()))
+      return 1;
+    size_t value = base_t::extents_type::static_extent(r);
+    return value == Kokkos::dynamic_extent ? 0 : value;
+  }
+};
+
+template <typename D, class... P>
+KOKKOS_INLINE_FUNCTION constexpr unsigned rank(const View<D, P...>&) {
+  return View<D, P...>::rank();
+}
+
+namespace Impl {
+
+template <typename ValueType, unsigned int Rank>
+struct RankDataType {
+  using type = typename RankDataType<ValueType, Rank - 1>::type*;
+};
+
+template <typename ValueType>
+struct RankDataType<ValueType, 0> {
+  using type = ValueType;
+};
+
+template <unsigned N, typename... Args>
+KOKKOS_FUNCTION std::enable_if_t<
+    N == View<Args...>::rank() &&
+        std::is_same_v<typename ViewTraits<Args...>::specialize, void>,
+    View<Args...> >
+as_view_of_rank_n(View<Args...> v) {
+  return v;
+}
+
+// Placeholder implementation to compile generic code for DynRankView; should
+// never be called
+template <unsigned N, typename T, typename... Args>
+KOKKOS_FUNCTION std::enable_if_t<
+    N != View<T, Args...>::rank() &&
+        std::is_same_v<typename ViewTraits<T, Args...>::specialize, void>,
+    View<typename RankDataType<typename View<T, Args...>::value_type, N>::type,
+         Args...> >
+as_view_of_rank_n(View<T, Args...>) {
+  Kokkos::abort("Trying to get at a View of the wrong rank");
```

Comment:
The abort message could be more expressive, but no hard feelings as this is for internal use only